### ISSUE TITLE
Change mlflow parameters for training

### ISSUE THREAD

**Issue by hrshovon** [ultralytics/ultralytics#20678]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,
I want to change mlflow parameters before training the model.
For example, I want to set a tag(such as train and validation etc.) because I am also doing a separate validation run to save the final **classwise** validation result in mlflow. 
For the validation run, I can set my own experiment as well as set tags because I am doing all that externally.
But for the training, I can only set the `project_name`. if I try to start training under my own mlflow run, I get errors like `connection reset by peers`.

Any tutorial/code snippet on how to get this done?

TIA

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hrshovon, thank you for reaching out and for your interest in Ultralytics 🚀! For questions about custom logging or MLflow integration, we recommend checking our [Docs](https://docs.ultralytics.com/), especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides, where many common workflows are covered.

If this is a 🐛 Bug Report, please help us by providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so that we can better understand and debug the issue.

For custom training ❓ questions like yours, please share as much detail as possible, including relevant code snippets, your training configuration, and any specific error messages. Also, check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for helpful advice.

Join the Ultralytics community for more support and discussion! Jump into real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧, explore in-depth topics on [Discourse](https://community.ultralytics.com/), or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Make sure you’re using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

_This is an automated response. An Ultralytics engineer will review your question and provide further assistance soon!_

**Comment by Y-T-G** [comment]
You can start an MLFlow run before running `model.train()`

**Comment by hrshovon** [comment]
> You can start an MLFlow run before running `model.train()`

I tried that, I am getting error like "connection reset by peers". could it be that the `train()` function is re-specifying something and in doing so creating some duplicate connection or something?

**Comment by glenn-jocher** [comment]
Based on the error you're encountering, it seems the Ultralytics MLflow integration is likely trying to create its own run and conflicting with your manually created one.

You can customize MLflow parameters before training without manually starting a run by using environment variables that the Ultralytics MLflow integration looks for:

```python
# Set MLflow parameters via environment variables
import os
os.environ["MLFLOW_EXPERIMENT_NAME"] = "your_experiment_name"
os.environ["MLFLOW_RUN"] = "your_run_name_with_tag" 
os.environ["MLFLOW_TRACKING_URI"] = "your_tracking_uri"

# Optional: keep the run active after training to add custom tags/metrics
os.environ["MLFLOW_KEEP_RUN_ACTIVE"] = "True"

# Train your model
model.train(data="data.yaml", epochs=10)

# If you set MLFLOW_KEEP_RUN_ACTIVE=True, you can add custom tags after training
import mlflow
mlflow.set_tag("run_type", "training")
mlflow.end_run()  # Don't forget to end the run when done
```

This approach lets you customize the MLflow parameters while letting Ultralytics handle the run lifecycle properly.

**Comment by hrshovon** [comment]
> Based on the error you're encountering, it seems the Ultralytics MLflow integration is likely trying to create its own run and conflicting with your manually created one.
> 
> You can customize MLflow parameters before training without manually starting a run by using environment variables that the Ultralytics MLflow integration looks for:
> 
> # Set MLflow parameters via environment variables
> import os
> os.environ["MLFLOW_EXPERIMENT_NAME"] = "your_experiment_name"
> os.environ["MLFLOW_RUN"] = "your_run_name_with_tag" 
> os.environ["MLFLOW_TRACKING_URI"] = "your_tracking_uri"
> 
> # Optional: keep the run active after training to add custom tags/metrics
> os.environ["MLFLOW_KEEP_RUN_ACTIVE"] = "True"
> 
> # Train your model
> model.train(data="data.yaml", epochs=10)
> 
> # If you set MLFLOW_KEEP_RUN_ACTIVE=True, you can add custom tags after training
> import mlflow
> mlflow.set_tag("run_type", "training")
> mlflow.end_run()  # Don't forget to end the run when done
> This approach lets you customize the MLflow parameters while letting Ultralytics handle the run lifecycle properly.

Code I am using
```
from ultralytics import YOLO 
import mlflow 
import os 
from pathlib import Path
import time 
mlflow.set_tracking_uri(xxxx)


experiment_name = "xxxxxx"
model_name = "yolo12l.pt"
epochs = 1
image_size=640
batch=32

general_params = {
    "half": True,
    "multi_scale": True,
    "save_period": 5
}

lr_params = {
    "cos_lr":True
}

augmentation_params = {
    "hsv_h":0.015,
    "hsv_s":0.7,
    "hsv_v":0.4,
    "translate":0.1,
    "scale":0.5,
    "fliplr":0.5,
    "mosaic":1.0,
    "erasing":0.4,
    "auto_augment":"randaugment"
}

all_params = dict(**general_params, **lr_params, **augmentation_params)


run_name = f"{model_name}-{epochs}-{image_size}-{time.time()}"

os.environ["MLFLOW_EXPERIMENT_NAME"] = experiment_name
os.environ["MLFLOW_RUN"] = run_name

# Optional: keep the run active after training to add custom tags/metrics
os.environ["MLFLOW_KEEP_RUN_ACTIVE"] = "True"

model = YOLO('yolo12l.pt')


model.train(data='dataset.yaml', epochs=epochs, device=[1,2], batch=batch, imgsz=image_size, **all_params)

mlflow.set_tag("task","train")
    
mlflow.end_run()

mlflow.set_experiment(f"{experiment_name}")

with mlflow.start_run(run_name=f"{run_name}"):    
    mlflow.set_tag("task","val")
    val_results = model.val()

    class_names = model.names.values()
    
    data_out = {f"classes/{classname}":map50 for i, (classname, map50) in enumerate(zip(class_names,val_results.box.ap50))}
    mlflow.log_metrics(data_out)
    mlflow.end_run()
    
```
Here i am not getting any errors, but the tag(`mlflow.set_tag("task","train")`) is being saved in a different random named run.

**Comment by glenn-jocher** [comment]
I see the issue now. When using `MLFLOW_KEEP_RUN_ACTIVE=True`, the run is kept active, but you're not capturing a reference to that active run before calling `mlflow.end_run()`.

Try this approach to get a reference to the MLflow run created during training:

```python
# Set MLflow parameters via environment variables
os.environ["MLFLOW_EXPERIMENT_NAME"] = experiment_name
os.environ["MLFLOW_RUN"] = run_name
os.environ["MLFLOW_KEEP_RUN_ACTIVE"] = "True"

model = YOLO('yolo12l.pt')
model.train(data='dataset.yaml', epochs=epochs, device=[1,2], batch=batch, imgsz=image_size, **all_params)

# Get the current active run (should be the one from training)
active_run = mlflow.active_run()
run_id = active_run.info.run_id

# Set tag on the training run
mlflow.set_tag("task", "train")
mlflow.end_run()

# For validation, start a new run with the same name but add a suffix
with mlflow.start_run(run_name=f"{run_name}-validation"):
    mlflow.set_tag("task", "val")
    val_results = model.val()
    
    class_names = model.names.values()
    data_out = {f"classes/{classname}": map50 for i, (classname, map50) in enumerate(zip(class_names, val_results.box.ap50))}
    mlflow.log_metrics(data_out)
```

This approach ensures you're working with the actual run that Ultralytics created during training before ending it.

**Comment by hrshovon** [comment]
> I see the issue now. When using `MLFLOW_KEEP_RUN_ACTIVE=True`, the run is kept active, but you're not capturing a reference to that active run before calling `mlflow.end_run()`.
> 
> Try this approach to get a reference to the MLflow run created during training:
> 
> # Set MLflow parameters via environment variables
> os.environ["MLFLOW_EXPERIMENT_NAME"] = experiment_name
> os.environ["MLFLOW_RUN"] = run_name
> os.environ["MLFLOW_KEEP_RUN_ACTIVE"] = "True"
> 
> model = YOLO('yolo12l.pt')
> model.train(data='dataset.yaml', epochs=epochs, device=[1,2], batch=batch, imgsz=image_size, **all_params)
> 
> # Get the current active run (should be the one from training)
> active_run = mlflow.active_run()
> run_id = active_run.info.run_id
> 
> # Set tag on the training run
> mlflow.set_tag("task", "train")
> mlflow.end_run()
> 
> # For validation, start a new run with the same name but add a suffix
> with mlflow.start_run(run_name=f"{run_name}-validation"):
>     mlflow.set_tag("task", "val")
>     val_results = model.val()
>     
>     class_names = model.names.values()
>     data_out = {f"classes/{classname}": map50 for i, (classname, map50) in enumerate(zip(class_names, val_results.box.ap50))}
>     mlflow.log_metrics(data_out)
> This approach ensures you're working with the actual run that Ultralytics created during training before ending it.

thanks for the quick response. How to use that `run_id` variable? I dont see it being referenced anywhere.
